Fix SILType::isEscapable for box types.

atrick opened the pull request:
SILBoxTypes have their own generic signature and substitution
map. This means that every time we query isEscapable or mayEscape, we
need to extract the type of the box's field and perform type
substitution so that the AST query only sees types from the function's
generic environment.

Fixes rdar://124179106 (Assertion failed in SIL:
(!type->hasTypeParameter() && "caller forgot to mapTypeIntoContext!"))

atrick commented:
@swift-ci test

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LifetimeDependenceDiagnostics.swift:
```diff
@@ -35,7 +35,8 @@ let lifetimeDependenceDiagnosticsPass = FunctionPass(
 
-  for argument in function.arguments where !argument.type.isEscapable {
+  for argument in function.arguments
+      where !argument.type.isEscapable(in: function) {
```
nitpick: for multi-line control statements the `{` should be in a new line

eeckstein reviewed (approved):
lgtm!

atrick commented:
@swift-ci smoke test

atrick committed:
Fix SILType::isEscapable for box types.

SILBoxTypes have their own generic signature and substitution
map. This means that every time we query isEscapable or mayEscape, we
need to extract the type of the box's field and perform type
substitution so that the AST query only sees types from the function's
generic environment.

Fixes rdar://124179106 (Assertion failed in SIL:
(!type->hasTypeParameter() && "caller forgot to mapTypeIntoContext!"))

atrick commented:
@swift-ci smoke test

atrick commented:
@swift-ci smoke test linux

atrick commented:
@swift-ci smoke test linux
